File: fase1-coleta/scraper.py
"""
OPER RADAR — Fase 1
Orquestrador principal. Pensado para rodar 2x/dia via cron (07h e 19h).

Uso: python scraper.py --janela 07h
"""
import argparse
import hashlib
import re
import sys
from datetime import datetime, timezone

import requests
import psycopg2
from psycopg2.extras import RealDictCursor

from parser import parse_listings, hash_pagina
from diff_logic import processa_diff, EstadoAnuncio

HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"}
BASE_URL = "https://www.caminhoesecarretas.com.br"
# Os links no HTML real são RELATIVOS (sem domínio, sem barra inicial) — confirmado via
# curl+grep direto no servidor de produção.
LOJA_URL_RE = re.compile(r'href="([a-z0-9-]+/[a-z]{2}/loja/[a-z0-9-]+/veiculo/\d+)"')


def discover_revenda_urls(uf: str) -> list[str]:
    """Varre revendas.aspx?uf=XX e devolve as URLs REAIS de perfil de cada revenda.
    Isso substitui qualquer URL "adivinhada" por slug — usa só o que o portal realmente linka."""
    resp = requests.get(f"https://www.caminhoesecarretas.com.br/revendas.aspx?uf={uf}", headers=HEADERS, timeout=20)
    resp.raise_for_status()
    caminhos = sorted(set(LOJA_URL_RE.findall(resp.text)))
    return [f"{BASE_URL}/{caminho}" for caminho in caminhos]


def fetch_revenda_page(url: str) -> str:
    resp = requests.get(url, headers=HEADERS, timeout=20)
    resp.raise_for_status()
    return resp.text


def get_or_create_revenda(conn, nome: str, cidade: str, uf: str, url_perfil: str) -> int:
    with conn.cursor() as cur:
        cur.execute("SELECT id FROM revenda WHERE url_perfil = %s", (url_perfil,))
        row = cur.fetchone()
        if row:
            return row[0]
        cur.execute(
            "INSERT INTO revenda (nome, cidade, uf, url_perfil) VALUES (%s,%s,%s,%s) RETURNING id",
            (nome, cidade, uf, url_perfil),
        )
        return cur.fetchone()[0]


def carrega_estado_atual(conn, revenda_id: int) -> dict[int, EstadoAnuncio]:
    estado = {}
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute(
            "SELECT anuncio_portal_id, status, misses_consecutivos, primeira_vez_visto, "
            "ultima_vez_ativo, data_remocao FROM anuncio WHERE revenda_id = %s", (revenda_id,)
        )
        for row in cur.fetchall():
            estado[row["anuncio_portal_id"]] = EstadoAnuncio(
                anuncio_portal_id=row["anuncio_portal_id"], status=row["status"],
                misses_consecutivos=row["misses_consecutivos"], primeira_vez_visto=row["primeira_vez_visto"],
                ultima_vez_ativo=row["ultima_vez_ativo"], data_remocao=row["data_remocao"],
            )
    return estado


def salva_estado(conn, revenda_id: int, novo_estado: dict, anuncios_por_id: dict):
    with conn.cursor() as cur:
        for anuncio_id, estado in novo_estado.items():
            anuncio = anuncios_por_id.get(anuncio_id)
            if anuncio is not None:
                # Anúncio visto nesta coleta (novo ou continuando ativo) — grava tudo.
                cur.execute("""
                    INSERT INTO anuncio (anuncio_portal_id, revenda_id, url, titulo, tipo, marca,
                        ano_inicial, ano_final, km_ou_horas, preco, preco_texto_bruto,
                        primeira_vez_visto, ultima_vez_ativo, status, misses_consecutivos, data_remocao)
                    VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                    ON CONFLICT (revenda_id, anuncio_portal_id) DO UPDATE SET
                        ultima_vez_ativo = EXCLUDED.ultima_vez_ativo,
                        status = EXCLUDED.status,
                        misses_consecutivos = EXCLUDED.misses_consecutivos,
                        data_remocao = EXCLUDED.data_remocao,
                        ano_inicial = COALESCE(EXCLUDED.ano_inicial, anuncio.ano_inicial),
                        ano_final = COALESCE(EXCLUDED.ano_final, anuncio.ano_final),
                        km_ou_horas = COALESCE(NULLIF(EXCLUDED.km_ou_horas, ''), anuncio.km_ou_horas),
                        preco = COALESCE(EXCLUDED.preco, anuncio.preco)
                """, (
                    anuncio_id, revenda_id,
                    anuncio.url, anuncio.titulo, anuncio.tipo, anuncio.marca,
                    anuncio.ano_inicial, anuncio.ano_final, anuncio.km_ou_horas,
                    anuncio.preco, anuncio.preco_texto_bruto,
                    estado.primeira_vez_visto, estado.ultima_vez_ativo, estado.status,
                    estado.misses_consecutivos, estado.data_remocao,
                ))
            else:
                # Anúncio que já existia e sumiu desta coleta — só atualiza o status.
                cur.execute("""
                    UPDATE anuncio SET
                        ultima_vez_ativo = %s, status = %s, misses_consecutivos = %s, data_remocao = %s
                    WHERE revenda_id = %s AND anuncio_portal_id = %s
                """, (
                    estado.ultima_vez_ativo, estado.status, estado.misses_consecutivos,
                    estado.data_remocao, revenda_id, anuncio_id,
                ))
    conn.commit()


def roda_ciclo(conn, uf: str, janela: str):
    urls = discover_revenda_urls(uf)
    print(f"[{uf}] {len(urls)} revendas encontradas")

    for url in urls:
        try:
            html = fetch_revenda_page(url)
        except requests.RequestException as e:
            registra_execucao(conn, None, janela, 0, "", sucesso=False, erro=str(e))
            continue

        h = hash_pagina(html)
        # (a lógica de "pular se hash igual" fica no caller, comparando com a última
        #  execucao_coleta dessa revenda antes de chamar fetch_revenda_page — aqui já
        #  buscamos, então logamos o hash normalmente)

        nome_revenda = url.split("/loja/")[1].split("/")[0].replace("%20", " ").replace("-", " ").title()
        cidade = url.split("/")[3].replace("-", " ").title()
        revenda_id = get_or_create_revenda(conn, nome_revenda, cidade, uf.upper(), url)

        anuncios = parse_listings(html)
        anuncios_por_id = {a.anuncio_portal_id: a for a in anuncios}
        ids_ativos = set(anuncios_por_id.keys())

        estado_anterior = carrega_estado_atual(conn, revenda_id)
        agora = datetime.now(timezone.utc)
        novo_estado = processa_diff(estado_anterior, ids_ativos, agora)
        salva_estado(conn, revenda_id, novo_estado, anuncios_por_id)

        registra_execucao(conn, revenda_id, janela, len(ids_ativos), h, sucesso=True)
        print(f"  {nome_revenda}: {len(ids_ativos)} anúncios ativos")


def registra_execucao(conn, revenda_id, janela, qtd_ativos, hash_pag, sucesso, erro=None):
    with conn.cursor() as cur:
        cur.execute("""
            INSERT INTO execucao_coleta (revenda_id, janela, qtd_anuncios_ativos, hash_pagina, sucesso, erro_mensagem)
            VALUES (%s,%s,%s,%s,%s,%s)
        """, (revenda_id, janela, qtd_ativos, hash_pag, sucesso, erro))
    conn.commit()


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("--janela", choices=["07h", "19h"], required=True)
    ap.add_argument("--uf", default="PR")
    ap.add_argument("--dsn", default="dbname=oper_radar user=oper_radar")
    args = ap.parse_args()

    conn = psycopg2.connect(args.dsn)
    try:
        roda_ciclo(conn, args.uf, args.janela)
    finally:
        conn.close()

File: fase1-coleta-php/scraper.php
<?php
/**
 * OPER RADAR — Fase 1 (versão PHP para HostGator Plano M, sem Python/terminal disponível)
 *
 * Usa cURL (já vem pronto no HostGator) e MySQLi (idem) — não precisa instalar nada.
 * Roda via Tarefas Cron do cPanel, lendo as credenciais de variaveis de ambiente.
 *
 * Uso: set -a; . /home/SEUUSUARIO/.oper-radar.env; set +a; php scraper.php --janela=07h --uf=PR
 */
require_once __DIR__ . '/parser.php';
require_once __DIR__ . '/diff_logic.php';

const USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';
const BASE_URL = 'https://www.caminhoesecarretas.com.br';
// Os links no HTML real sao RELATIVOS (sem dominio, sem barra inicial) — confirmado
// via curl+grep direto no servidor de producao.
const LOJA_URL_RE = '/href="([a-z0-9-]+\/[a-z]{2}\/loja\/[a-z0-9-]+\/veiculo\/\d+)"/';

// Pausa entre cada revenda visitada. Plano compartilhado tem limite de 25% de CPU por
// períodos >=90s — rodar devagar evita estourar isso (e é mais gentil com o portal).
const PAUSA_ENTRE_REQUISICOES_SEGUNDOS = 2;

function http_get(string $url): string {
    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_USERAGENT => USER_AGENT,
        CURLOPT_TIMEOUT => 20,
        CURLOPT_FOLLOWLOCATION => true,
    ]);
    $resp = curl_exec($ch);
    if ($resp === false) {
        throw new RuntimeException('cURL error: ' . curl_error($ch));
    }
    curl_close($ch);
    return $resp;
}

function discover_revenda_urls(string $uf): array {
    $html = http_get("https://www.caminhoesecarretas.com.br/revendas.aspx?uf={$uf}");
    preg_match_all(LOJA_URL_RE, $html, $matches);
    $caminhos = array_values(array_unique($matches[1]));
    return array_map(fn($c) => BASE_URL . '/' . $c, $caminhos);
}

function get_or_create_revenda(mysqli $conn, string $nome, string $cidade, string $uf, string $url_perfil): int {
    $stmt = $conn->prepare('SELECT id FROM revenda WHERE url_perfil = ?');
    $stmt->bind_param('s', $url_perfil);
    $stmt->execute();
    $res = $stmt->get_result()->fetch_assoc();
    $stmt->close();
    if ($res) {
        return (int) $res['id'];
    }
    $stmt = $conn->prepare('INSERT INTO revenda (nome, cidade, uf, url_perfil) VALUES (?,?,?,?)');
    $stmt->bind_param('ssss', $nome, $cidade, $uf, $url_perfil);
    $stmt->execute();
    $novo_id = $conn->insert_id;
    $stmt->close();
    return $novo_id;
}

function carrega_estado_atual(mysqli $conn, int $revenda_id): array {
    $estado = [];
    $stmt = $conn->prepare(
        'SELECT anuncio_portal_id, status, misses_consecutivos, primeira_vez_visto, ' .
        'ultima_vez_ativo, data_remocao FROM anuncio WHERE revenda_id = ?'
    );
    $stmt->bind_param('i', $revenda_id);
    $stmt->execute();
    $res = $stmt->get_result();
    while ($row = $res->fetch_assoc()) {
        $estado[(int) $row['anuncio_portal_id']] = [
            'status' => $row['status'],
            'misses_consecutivos' => (int) $row['misses_consecutivos'],
            'primeira_vez_visto' => $row['primeira_vez_visto'],
            'ultima_vez_ativo' => $row['ultima_vez_ativo'],
            'data_remocao' => $row['data_remocao'],
        ];
    }
    $stmt->close();
    return $estado;
}

function salva_estado(mysqli $conn, int $revenda_id, array $novo_estado, array $anuncios_por_id): void {
    $sqlInsert = "INSERT INTO anuncio (anuncio_portal_id, revenda_id, url, titulo, tipo, marca,
                ano_inicial, ano_final, km_ou_horas, preco, preco_texto_bruto,
                primeira_vez_visto, ultima_vez_ativo, status, misses_consecutivos, data_remocao)
            VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
            ON DUPLICATE KEY UPDATE
                ultima_vez_ativo = VALUES(ultima_vez_ativo),
                status = VALUES(status),
                misses_consecutivos = VALUES(misses_consecutivos),
                data_remocao = VALUES(data_remocao),
                ano_inicial = COALESCE(VALUES(ano_inicial), ano_inicial),
                ano_final = COALESCE(VALUES(ano_final), ano_final),
                km_ou_horas = COALESCE(NULLIF(VALUES(km_ou_horas), ''), km_ou_horas),
                preco = COALESCE(VALUES(preco), preco)";
    $stmtInsert = $conn->prepare($sqlInsert);

    $sqlUpdateOnly = "UPDATE anuncio SET
                ultima_vez_ativo = ?, status = ?, misses_consecutivos = ?, data_remocao = ?
            WHERE revenda_id = ? AND anuncio_portal_id = ?";
    $stmtUpdate = $conn->prepare($sqlUpdateOnly);

    foreach ($novo_estado as $anuncio_id => $estado) {
        $a = $anuncios_por_id[$anuncio_id] ?? null;

        if ($a !== null) {
            // Anúncio visto nesta coleta — grava tudo.
            $url = $a['url'];
            $titulo = $a['titulo'];
            $tipo = $a['tipo'];
            $marca = $a['marca'];
            $anoInicial = $a['ano_inicial'];
            $anoFinal = $a['ano_final'];
            $preco = $a['preco'];
            $precoTexto = $a['preco_texto_bruto'];
            $kmOuHoras = $a['km_ou_horas'];

            $stmtInsert->bind_param(
                'iissssiisdssssis',
                $anuncio_id, $revenda_id, $url, $titulo, $tipo, $marca,
                $anoInicial, $anoFinal, $kmOuHoras, $preco, $precoTexto,
                $estado['primeira_vez_visto'], $estado['ultima_vez_ativo'], $estado['status'],
                $estado['misses_consecutivos'], $estado['data_remocao']
            );
            $stmtInsert->execute();
        } else {
            // Anúncio que já existia e sumiu desta coleta — só atualiza o status,
            // não mexe em url/titulo/preço (evita violar NOT NULL).
            $stmtUpdate->bind_param(
                'ssisi',
                $estado['ultima_vez_ativo'], $estado['status'], $estado['misses_consecutivos'],
                $estado['data_remocao'], $revenda_id, $anuncio_id
            );
            $stmtUpdate->execute();
        }
    }
    $stmtInsert->close();
    $stmtUpdate->close();
}

function registra_execucao(mysqli $conn, ?int $revenda_id, string $janela, int $qtd_ativos, string $hash, bool $sucesso, ?string $erro = null): void {
    $stmt = $conn->prepare(
        'INSERT INTO execucao_coleta (revenda_id, janela, qtd_anuncios_ativos, hash_pagina, sucesso, erro_mensagem) VALUES (?,?,?,?,?,?)'
    );
    $stmt->bind_param('isissi', $revenda_id, $janela, $qtd_ativos, $hash, $sucesso, $erro);
    $stmt->execute();
    $stmt->close();
}

function roda_ciclo(mysqli $conn, string $uf, string $janela, array $marcas_conhecidas): void {
    $urls = discover_revenda_urls($uf);
    $total = count($urls);
    echo "[{$uf}] {$total} revendas encontradas — pausa de " . PAUSA_ENTRE_REQUISICOES_SEGUNDOS . "s entre cada uma\n";

    foreach ($urls as $i => $url) {
        try {
            $html = http_get($url);
        } catch (RuntimeException $e) {
            registra_execucao($conn, null, $janela, 0, '', false, $e->getMessage());
            sleep(PAUSA_ENTRE_REQUISICOES_SEGUNDOS);
            continue;
        }

        $hash = hash_pagina($html);
        $partesLoja = explode('/loja/', $url);
        $nomeRevenda = ucwords(str_replace(['%20', '-'], [' ', ' '], explode('/', $partesLoja[1])[0]));
        $cidade = ucwords(str_replace('-', ' ', explode('/', $url)[3]));
        $revendaId = get_or_create_revenda($conn, $nomeRevenda, $cidade, strtoupper($uf), $url);

        $anuncios = parse_listings($html, $marcas_conhecidas);
        $anunciosPorId = [];
        foreach ($anuncios as $a) {
            $anunciosPorId[$a['anuncio_portal_id']] = $a;
        }
        $idsAtivos = array_keys($anunciosPorId);

        $estadoAnterior = carrega_estado_atual($conn, $revendaId);
        $agora = date('Y-m-d H:i:s');
        $novoEstado = processa_diff($estadoAnterior, $idsAtivos, $agora);
        salva_estado($conn, $revendaId, $novoEstado, $anunciosPorId);

        registra_execucao($conn, $revendaId, $janela, count($idsAtivos), $hash, true);
        printf("  [%d/%d] %s: %d anúncios ativos\n", $i + 1, $total, $nomeRevenda, count($idsAtivos));

        if ($i < $total - 1) {
            sleep(PAUSA_ENTRE_REQUISICOES_SEGUNDOS);
        }
    }
}

// ---- CLI entry point ----
if (php_sapi_name() === 'cli' && basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'] ?? '')) {
    $opts = getopt('', ['janela:', 'uf::', 'db-host::', 'db-user::', 'db-pass::', 'db-name::']);

    $janela = $opts['janela'] ?? null;
    if (!in_array($janela, ['07h', '19h'], true)) {
        fwrite(STDERR, "Uso: php scraper.php --janela=07h|19h --uf=PR (credenciais via OPER_RADAR_DB_*)\n");
        exit(1);
    }
    $uf = $opts['uf'] ?? 'PR';
    $dbHost = $opts['db-host'] ?? (getenv('OPER_RADAR_DB_HOST') ?: 'localhost');
    $dbUser = $opts['db-user'] ?? (getenv('OPER_RADAR_DB_USER') ?: null);
    $dbPass = $opts['db-pass'] ?? (getenv('OPER_RADAR_DB_PASS') ?: null);
    $dbName = $opts['db-name'] ?? (getenv('OPER_RADAR_DB_NAME') ?: null);

    if (!$dbUser || !$dbPass || !$dbName) {
        fwrite(STDERR, "Faltam OPER_RADAR_DB_USER, OPER_RADAR_DB_PASS ou OPER_RADAR_DB_NAME\n");
        exit(1);
    }

    $conn = new mysqli($dbHost, $dbUser, $dbPass, $dbName);
    if ($conn->connect_error) {
        fwrite(STDERR, "Erro de conexão MySQL: {$conn->connect_error}\n");
        exit(1);
    }
    $conn->set_charset('utf8mb4');

    $marcas = carrega_marcas_conhecidas();
    roda_ciclo($conn, $uf, $janela, $marcas);
    $conn->close();
}
